### PR TITLE
Expire

### PR DESCRIPTION
**Pull request by AlexeyVavilin**
ABI v2 with message expiration supported
Initial contract data on deploy

### CODE REVIEW

File: src/contracts.rs [hunk 1]
```diff
@@ -21,25 +21,45 @@ use crate::interop::{InteropContext, Interop};
 #[allow(non_snake_case)]
 pub(crate) struct ParamsOfDeploy {
     pub abi: serde_json::Value,
+    pub constructorHeader: Option<serde_json::Value>,
```
Comment (Keshoid) [inline src/contracts.rs:24]: Why it is constructorHeader?

File: src/contracts.rs [hunk 2]
```diff
@@ -21,25 +21,45 @@ use crate::interop::{InteropContext, Interop};
...
+pub(crate) struct ResultOfGetDeployDataCore {
+    pub imageBase64: Option<String>,
+    pub address: Option<String>,
+    pub dataBase64: String,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ResultOfGetDeployData {
+    pub image: Option<Vec<u8>>,
+    pub address: Option<TonAddress>,
```
Comment (Keshoid) [inline src/contracts.rs:61]: Why it is an `Option<address>`? When it can be empty?

File: src/contracts.rs [hunk 3]
```diff
@@ -109,44 +131,86 @@ impl TonContracts {
...
+        &self,
+        abi: Option<&str>,
+        image: Option<&[u8]>,
+        init_params: Option<RunParameters>,
+        public_key: &Ed25519Public,
+        workchain_id: Option<i32>,
+    ) -> TonResult<ResultOfGetDeployData> {
+        let abi = abi.map(|val| 
+                serde_json::from_str(val).map_err(|_| TonErrorKind::InvalidArg(val.to_owned())))
+            .transpose()?;
```
Comment (Keshoid) [inline src/contracts.rs:156]: Maybe format like this?
```
let abi = abi.map(|val| {
    serde_json::from_str(val).map_err(|_| TonErrorKind::InvalidArg(val.to_owned()))
}).transpose()?;
```
Or it is not a rust style)

File: src/contracts.rs [hunk 4]
```diff
@@ -109,44 +131,86 @@ impl TonContracts {
...
+            None => Ok(None)
+        }
     }
 
     /// Deploy contract to TON blockchain
     pub fn deploy(
         &self,
         abi: &str,
         code: &[u8],
+        constructor_header: Option<RunParameters>,
```
Comment (Keshoid) [inline src/contracts.rs:196]: Why not just a `header`? or `abi_header`?

File: src/contracts.rs [hunk 5]
```diff
@@ -155,26 +219,26 @@ impl TonContracts {
...
-            RunParameters::Json(string) => string
-        };
-        let params_value = serde_json::from_str(str_params)
-            .map_err(|_| TonErrorKind::InvalidArg(str_params.to_owned()))?;
 
-        let result: ResultOfRun = Interop::json_request(self.context, "contracts.run", ParamsOfRun {
+        let header = match header {
+            Some(header) => Some(Self::params_to_value(header)?),
+            None => None
+        };
```
Comment (Keshoid) [inline src/contracts.rs:232]: can you use  `transpose`  here?

File: src/contracts.rs (hunk 1 above)
Comment (AlexeyVavilin) [reply src/contracts.rs:24]: It contains header params for constructor call. Like `constructorParams` contains parameters for constructor itself

File: src/contracts.rs (hunk 2 above)
Comment (AlexeyVavilin) [reply src/contracts.rs:61]: It can be empty if `image` is not provided. We can't calculate address using only data

File: src/contracts.rs (hunk 3 above)
Comment (AlexeyVavilin) [reply src/contracts.rs:156]: Done

File: src/contracts.rs (hunk 4 above)
Comment (AlexeyVavilin) [reply src/contracts.rs:196]: It contains header params for constructor call. Like `constructor_params` contains parameters for constructor itself

File: src/contracts.rs (hunk 5 above)
Comment (AlexeyVavilin) [reply src/contracts.rs:232]: Done

File: src/client.rs [hunk 6]
```diff
@@ -31,6 +31,13 @@ use crate::{TonCrypto, TonContracts, TonQueries, TonResult};
 #[serde(rename_all="camelCase")]
 pub struct TonClientConfig {
     pub base_url: Option<String>,
+    pub message_retries_count: Option<u8>,
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/client.rs:34]: Done

**Pull request merged**